File: 117-populating-next-right-pointers-in-each-node-ii/populating-next-right-pointers-in-each-node-ii.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
}
*/

public class Solution {
    public Node Connect(Node root) {
        if (root == null) return null;
        
        Node leftmost = root;

        while (leftmost != null) {
            Node head = leftmost;
            Node prev = null;
            leftmost = null;

            while (head != null) {
                if (head.left != null) {
                    if (prev != null) {
                        prev.next = head.left;
                    } else {
                        leftmost = head.left;
                    }
                    prev = head.left;
                }

                if (head.right != null) {
                    if (prev != null) {
                        prev.next = head.right;
                    } else {
                        leftmost = head.right;
                    }
                    prev = head.right;
                }

                head = head.next;
            }
        }

        return root;
    }
}

File: 117-populating-next-right-pointers-in-each-node-ii/populating-next-right-pointers-in-each-node-ii.js
/**
 * // Definition for a _Node.
 * function _Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */

/**
 * @param {_Node} root
 * @return {_Node}
 */
var connect = function(root) {
    if (!root) return null;
    let queue = [root]; 
    while (queue.length > 0) {
        let levelSize = queue.length;
        let prevNode = null;
        for (let i = 0; i < levelSize; i++) {
            let currentNode = queue.shift();
            if (prevNode) {
                prevNode.next = currentNode;
            }
            prevNode = currentNode;
            if (currentNode.left) queue.push(currentNode.left);
            if (currentNode.right) queue.push(currentNode.right);
        }
        prevNode.next = null;
    }
    return root;
};
